x/ibc-transfer/keeper/relay.go

```diff
@@ -59,69 +61,54 @@ func (k Keeper) createOutgoingPacket(
 	seq uint64,
 	sourcePort, sourceChannel,
 	destinationPort, destinationChannel string,
-	amount sdk.Coins,
+	amount sdk.Coin,
 	sender sdk.AccAddress,
 	receiver string,
+	source bool,
 	timeoutHeight, timeoutTimestamp uint64,
 ) error {
 	channelCap, ok := k.scopedKeeper.GetCapability(ctx, host.ChannelCapabilityPath(sourcePort, sourceChannel))
 	if !ok {
 		return sdkerrors.Wrap(channeltypes.ErrChannelCapabilityNotFound, "module does not own channel capability")
 	}
-	// NOTE:
-	// - Coins transferred from the destination chain should have their denomination
-	// prefixed with source port and channel IDs.
-	// - Coins transferred from the source chain can have their denomination
-	// clear from prefixes when transferred to the escrow account (i.e when they are
-	// locked) BUT MUST have the destination port and channel ID when constructing
-	// the packet data.
-	if len(amount) != 1 {
-		return sdkerrors.Wrapf(types.ErrOnlyOneDenomAllowed, "%d denoms included", len(amount))
-	}
-
-	prefix := types.GetDenomPrefix(destinationPort, destinationChannel)
-	source := strings.HasPrefix(amount[0].Denom, prefix)
+	// NOTE: When the sender chain is acting as the source zone, the coins will
+	// be constructed with the destination port and channel ID prefixed when
+	// represented in the fungible token packet data. When being escrowed these
+	// coins will not have the prefix.
 
 	if source {
-		// clear the denomination from the prefix to send the coins to the escrow account
-		coins := make(sdk.Coins, len(amount))
-		for i, coin := range amount {
-			if strings.HasPrefix(coin.Denom, prefix) {
-				coins[i] = sdk.NewCoin(coin.Denom[len(prefix):], coin.Amount)
-			} else {
-				coins[i] = coin
-			}
-		}
-
-		// escrow tokens if the destination chain is the same as the sender's
+		// create the escrow address for the tokens
 		escrowAddress := types.GetEscrowAddress(sourcePort, sourceChannel)
 
 		// escrow source tokens. It fails if balance insufficient.
 		if err := k.bankKeeper.SendCoins(
-			ctx, sender, escrowAddress, coins,
+			ctx, sender, escrowAddress, sdk.NewCoins(amount),
 		); err != nil {
 			return err
 		}
 
+		// construct denom with prefix that will be used in the transfer
+		amount.Denom = types.GetPrefixedDenom(destinationPort, destinationChannel, amount.Denom)
+
```

Comment:
Why is the prefixing happening here? It makes more sense to me that sending chain just sets the source boolean, and the receiving chain prefixes or unprefixes the denom accordingly

Reply:
this is how the it is specified in the spec, but I also believe makes sense from a UI standpoint. The fungible token packet data must be consistent upon sending, receiving, and acknowledging. Not including the prefix in the send would lead to a receive packet/acknowledge packet being done with packet data that does not contain the prefix. I'm not sure which is better, but this would lead to a spec change so I'll differ to @cwgoes

Reply:
this is something that confuses me as well. It's weird that there's no prefix addition at all and that it has to be done on the client-side, which may not be so intuitive for relayers. The current spec only removes the prefix from the denomination in order to check correctness

Reply:
it should be noted this pr differs with the spec in that regard. client-side code is no longer responsible for prefixing on sends. On receives/acks it would need to know the prefix. So the flow looks like this:
```
client calls send of `atom` to portid and channelid with source to true
send  escrows atoms and creates packet with token denom `portid/channelid/atom`

client receives token `portid/channelid/atom` on counterparty with a fungible token packet using that denom

etc for ack on sender chain
```

Reply:
We could change this on the spec side, but I think it should be consistent across send/receive/acknowledgement

Would it be clearer just to have the base denomination in the packet / ack along with the `source` flag?

Reply:
Yes that would be clearer to me. We shouldn't trust the sender chain to prefix correctly, so we should be checking anyway. Given we have to do that, it's clearer to just prefix the denom on the receiver chain